hibernate-core/src/main/java/org/hibernate/hql/spi/id/subselect/inline/InlineIdsOrBuilder.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later.
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.hql.spi.id.subselect.inline;
+
+import java.util.List;
+
+/**
+ * @author Vlad Mihalcea
+ */
+public class InlineIdsOrBuilder extends IdsClauseBuilder {
+
+	private final String[] columns;
+
+	private StringBuilder statementBuilder;
+
+	public InlineIdsOrBuilder(
+			List<Object[]> ids,
+			String[] columns) {
+		super(ids);
+		this.columns = columns;
+
+		this.statementBuilder = buildStatement();
+	}
+
+	public String toStatement() {
+		return statementBuilder.toString();
+	}
+
+	private StringBuilder buildStatement() {
+		StringBuilder buffer = new StringBuilder();
+
+		String columnNames = String.join(",", columns);
+
+		for ( int i = 0; i < getIds().size(); i++ ) {
+			Object[] idTokens = getIds().get( i );
+			if ( i > 0 ) {
+				buffer.append( " or " );
+			}
+			buffer.append( columnNames );
+			buffer.append( " = ( " );
+			for ( int j = 0; j < idTokens.length; j++ ) {
+				if( j > 0) {
+					buffer.append( "," );
+				}
+				buffer.append( idTokens[j] );
+			}
+			buffer.append( ")" );
+		}
+
+		return buffer;
```

Comment:
If you know a limit of how many values are allowed in an `IN` clause (like 1000 for Oracle), instead of:
```sql
where COLUMN = 1 or COLUMN = 2 or COLUMN = 3 or COLUMN = 4 or COLUMN = 5 ...
   or COLUMN = 1001 or COLUMN = 1002 or ...
```
you could do something like this:
```sql
where COLUMN in (1, 2, 3, 4, 5 ...)
   or COLUMN in (1001, 1002, 1003, ...)
```
This way generated SQL is much shorter.